### PR TITLE
Add Tatoeba QnA dataset

### PR DESCRIPTION
**Pull request by 0x22almostEvil**
Multilingual Tatoeba Q&A Translation Dataset
============================================

120K entries

This dataset contains a list of instructions to translate or paraphrase in multiple languages. It is available in Parquet format and includes the following columns:

* INSTRUCTION: The instruction of text to be translated or paraphrased.
* RESPONSE: The corresponding response or answer in target language.
* SOURCE (tatoeba): The original source text from the Tatoeba database.
* METADATA (json): Additional information about each entry, including the target language, UUID, and pair of languages (source and target): `"{"language": "lang", "length": "length of original text","uuid": "uuid (original text + translated text)", "langs-pair": "from_lang-to_lang"}"`

The data in this dataset was collected through crowdsourcing efforts and includes translations of various types of content, such as sentences, phrases, idioms, and proverbs.

You can find it here: https://huggingface.co/datasets/0x22almostEvil/tatoeba-mt-qna-oa
Original dataset is available here: https://huggingface.co/datasets/Helsinki-NLP/tatoeba_mt

---

Resolves https://github.com/LAION-AI/Open-Assistant/issues/3115

### CODE REVIEW

File: data/datasets/tatoeba_mt_qna_oa/README.MD [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  translated text)", "langs-pair": "from_lang-to_lang"}"
+
+The data in this dataset was collected through crowdsourcing efforts and
+includes translations of various types of content, such as sentences, phrases,
+idioms, and proverbs.
+
+You can find it here:
+https://huggingface.co/datasets/0x22almostEvil/tatoeba-mt-qna-oa Original
+dataset is available here:
+https://huggingface.co/datasets/Helsinki-NLP/tatoeba_mt
```
Comment (sedthh) [inline data/datasets/tatoeba_mt_qna_oa/README.MD:24]: Please add a reference to the original HF dataset in your HF datacard as well.

File: data/datasets/tatoeba_mt_qna_oa/data_process.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+
+converter = DataProcess()
+
+"""
+EXAMPLES:
+
+# get language name; iso639_1 code
+print(converter.language_name('ru', 'en')) # Output: Russian
+print(converter.convert_code("eng")) # Output: en
```
Comment (sedthh) [inline data/datasets/tatoeba_mt_qna_oa/data_process.py:64]: "en" instead of "eng" ?

File: data/datasets/tatoeba_mt_qna_oa/data_process.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        "length": f"{text_length}",
+        "uuid": f"{uuid_val}",
+        "langs-pair": f"{lang_from}-{lang_to}",
+    }
+    metadata_str = json.dumps(METADATA)
+    source = "tatoeba"
+    # randomizing INSTRUCTION
+    instruction = converter.RandomText.randomize_text(text, lang_to, lang_from)
+    response = translation
+    return QnA(instruction, response, source, metadata_str)
```
Comment (sedthh) [inline data/datasets/tatoeba_mt_qna_oa/data_process.py:102]: Nice solution!

File: data/datasets/tatoeba_mt_qna_oa/data_process.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+hf_dataset = datasets.load_dataset("0x22almostEvil/tatoeba-mt-llama-only", split="train")
+
+# original is ~3M; with num_shards=30 it'll be ~120K
+hf_dataset = hf_dataset.shard(num_shards=30, index=0)
+print(hf_dataset)
+
+# convert the dataset to a pandas dataframe
+df = pd.DataFrame(hf_dataset)
+
+# apply the create_qna function to each row of the dataframe to create QnA objects
```
Comment (sedthh) [inline data/datasets/tatoeba_mt_qna_oa/data_process.py:115]: you can add a deterministic random seed here, so every run would yield the exact same randomly selected QnAs`
random.seed(1234567)`

File: data/datasets/tatoeba_mt_qna_oa/language_names.py [hunk 5]
```diff
@@ -0,0 +1,442 @@
+language_names = {
```
Comment (sedthh) [inline data/datasets/tatoeba_mt_qna_oa/language_names.py:1]: Nice, comprehensive list!

File: data/datasets/tatoeba_mt_qna_oa/data_process.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+import datasets
+import iso639
+import language_names
+import language_paraphrase
+import language_translate
+import pandas as pd
+
+
+class DataProcess:
+    class RandomText:
```
[the remark this answers is not in the document]
Comment (0x22almostEvil) [reply data/datasets/tatoeba_mt_qna_oa/data_process.py:15]: Nope, gonna fix 😅

File: data/datasets/tatoeba_mt_qna_oa/data_process.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        # provide instruction with a text; process of randomization of a text
+        @staticmethod
+        def randomize_text(text, original_lang=None, target_lang=None):
+            templates = (
+                language_translate.random_templates_translate.get(original_lang, {})
+                if not ((original_lang == target_lang) and (original_lang is not None) and (target_lang is not None))
+                else language_paraphrase.random_templates_paraphrase.get(original_lang, {})
+            )
+            template = random.choice(list(templates.values()))
+            quote = random.choice(list(DataProcess.RandomText().random_quote.values()))
```
[the remark this answers is not in the document]
Comment (0x22almostEvil) [reply data/datasets/tatoeba_mt_qna_oa/data_process.py:34]: I thought about this, can do it for sure

File: data/datasets/tatoeba_mt_qna_oa/README.MD (hunk 1 above)
Comment (0x22almostEvil) [reply data/datasets/tatoeba_mt_qna_oa/README.MD:24]: Done

File: data/datasets/tatoeba_mt_qna_oa/language_translate.py [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+        5: "Molim te, prevedi {quote}{text}{quote} na {lang2}",
+        6: "Kako se kaže {quote}{text}{quote}?",
+        7: "Tražim prijevod za {quote}{text}{quote} na {lang2}",
+        8: "Jesi li u mogućnosti prevesti {quote}{text}{quote} na {lang2}?",
+        9: "Koji je točan prijevod za {quote}{text}{quote} na {lang2}?",
+        10: "Možeš li mi dati prijevod za {quote}{text}{quote} na {lang2}?",
+        11: "{lang1}-{lang2}: {quote}{text}{quote}",
+    },
+    "hu": {
+        1: "Fordítsd le {quote}{text}{quote} {lang2}-ra innen: {lang1}",
```
[the remark this answers is not in the document]
Comment (sedthh) [reply data/datasets/tatoeba_mt_qna_oa/language_translate.py:145]:
```python
"hu": {
   1: "Fordítsd le magyar nyelvről {lang2} nyelvre: {quote}{text}{quote}",
   2: "Hogy mondanád {lang2} nyelven, hogy {quote}{text}{quote}?",
   3: "Kérlek fordítsd le magyarra: {text}",
   4: "Mit jelent, hogy {quote}{text}{quote}? Fordítsd le légy szíves!",
   5: "Hogyan mondod magyarul, hogy {quote}{text}{quote}?",
   6: "Fordítsd magyarra {lang1} nyelvről: {text}",
   7: "Mit jelent magyarul az a kifejezés, hogy {quote}{text}{quote}?",
   8: "Mondd magyarul: {quote}{text}{quote}",
   9: "Fordítsd le, hogy {quote}{text}{quote}",
   10: "Mondd magyarul, hogy {quote}{text}{quote}",
   11: "Mit jelent {lang2} nyelven: {text}",
   12: "{lang1}-{lang2}: {quote}{text}{quote}",
}

```

File: data/datasets/tatoeba_mt_qna_oa/data_process.py (hunk 2 above)
Comment (0x22almostEvil) [reply data/datasets/tatoeba_mt_qna_oa/data_process.py:64]: (Discussed in DM)

**Comment by 0x22almostEvil** [comment]
Not ready atm, gonna work on this for other langs too:
image

**Review by sedthh — APPROVED** [review]
LGTM

**Pull request merged**